wren-ui/src/apollo/server/services/sqlPairService.ts

```diff
@@ -0,0 +1,246 @@
+      projectId,
+    });
+    if (!existingPair) {
+      throw new Error(
+        `SQL pair with ID ${sqlPairId} not found in project ${projectId}`,
+      );
+    }
+
+    // Update only the provided fields
+    const updatedData: Partial<SqlPair> = {
+      sql: existingPair.sql,
+      question: existingPair.question,
+    };
+
+    if (sqlPair.sql !== undefined) {
+      updatedData.sql = sqlPair.sql;
+    }
+
+    if (sqlPair.question !== undefined) {
+      updatedData.question = sqlPair.question;
+    }
+    const tx = await this.sqlPairRepository.transaction();
+    try {
+      const updatedSqlPair = await this.sqlPairRepository.updateOne(
+        sqlPairId,
+        updatedData,
+        { tx },
+      );
+      const { queryId } = await this.wrenAIAdaptor.deploySqlPair(
+        projectId,
+        updatedSqlPair,
+      );
+      const deployResult = await this.waitUntilSqlPairResult(queryId);
+      if (deployResult.error) {
+        throw Errors.create(Errors.GeneralErrorCodes.DEPLOY_SQL_PAIR_ERROR, {
+          customMessage: deployResult.error.message,
+        });
+      }
+      await tx.commit();
+      return updatedSqlPair;
+    } catch (error) {
+      logger.error(`edit sql pair failed. ${error}`);
+      await tx.rollback();
+      throw Errors.create(Errors.GeneralErrorCodes.DEPLOY_SQL_PAIR_ERROR, {
+        customMessage: error.message,
+      });
+    }
+  }
+
+  async deleteSqlPair(projectId: number, sqlPairId: number): Promise<boolean> {
+    // First verify the SQL pair exists and belongs to the project
+    const existingPair = await this.sqlPairRepository.findOneBy({
+      id: sqlPairId,
+      projectId,
+    });
+
+    if (!existingPair) {
+      throw new Error(
+        `SQL pair with ID ${sqlPairId} not found in project ${projectId}`,
+      );
+    }
+    const tx = await this.sqlPairRepository.transaction();
+    try {
+      await this.sqlPairRepository.deleteOne(sqlPairId, { tx });
+      await this.wrenAIAdaptor.deleteSqlPairs(projectId, [sqlPairId]);
+      await tx.commit();
+      return true;
+    } catch (error) {
+      logger.error(`delete sql pair failed. ${error}`);
+      await tx.rollback();
+      throw Errors.create(Errors.GeneralErrorCodes.DEPLOY_SQL_PAIR_ERROR, {
+        customMessage: error.message,
+      });
+    }
+  }
+
+  private async waitUntilSqlPairResult(
+    queryId: string,
+  ): Promise<SqlPairResult> {
+    let result = await this.wrenAIAdaptor.getSqlPairResult(queryId);
+    while (!this.isFinishedState(result.status)) {
+      await new Promise((resolve) => setTimeout(resolve, 500));
+      result = await this.wrenAIAdaptor.getSqlPairResult(queryId);
+    }
+    return result;
+  }
+
+  private async waitQuestionGenerateResult(
+    queryId: string,
+  ): Promise<Partial<QuestionsResult>> {
+    let result = await this.wrenAIAdaptor.getQuestionsResult(queryId);
+    while (
+      ![QuestionsStatus.SUCCEEDED, SqlPairStatus.FAILED].includes(result.status)
```

Comment:
Maybe use `QuestionsStatus` instead ?
```suggestion
      ![QuestionsStatus.SUCCEEDED, QuestionsStatus.FAILED].includes(result.status)
```

Reply:
delivered in https://github.com/Canner/WrenAI/pull/1372/commits/42419ea531a7c181b35edf972536ea9e99b4520a